### PR TITLE
Implement cache for registry

### PR DESCRIPTION
**Pull request by lobkovilya**
Signed-off-by: Ilya Lobkov <lobkovilya@yandex.ru>

Implement caches for the following custom resources: `NetworkService`, `NetworkServiceEndpoint`, `NetworkServiceManager`. Each resource cache listens k8s-registry using `Informer` as well as handles `Add(...)` or `Delete(...)` requests from `nsmd-k8s` 

## How Has This Been Tested?
* [x]  Covered by existing integration testing
* [ ]  Added integration testing to cover
* [x]  Tested locally
* [ ]  Have not tested

## Types of changes
* [ ]  Bug fix (non-breaking change which fixes an issue)
* [ ]  New feature (non-breaking change which adds functionality)
* [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
* [ ]  My change requires a change to the documentation.
* [ ]  I have updated the documentation accordingly.

### CODE REVIEW

**Comment by lobkovilya** [comment]
@nickolaev @edwarnicke please take a look :)

File: test/integration/nsmd_test_utils/test_utils.go [hunk 1]
```diff
@@ -55,18 +55,18 @@ func deployNsmdAndDataplane(k8s *kube_testing.K8s, node *v1.Node) (nsmd *v1.Pod,
 	return
 }
 
-func DeployIcmp(k8s *kube_testing.K8s, node *v1.Node) (icmp *v1.Pod) {
+func DeployIcmp(k8s *kube_testing.K8s, node *v1.Node, name string) (icmp *v1.Pod) {
```
Comment (nickolaev) [inline test/integration/nsmd_test_utils/test_utils.go]: Can you please split this change in another patch and explain why we need it?

File: k8s/pkg/registryserver/registry_cache.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+func (rc *registryCacheImpl) GetNetworkService(name string) (*v1.NetworkService, error) {
+	if ns := rc.networkServiceCache.Get(name); ns == nil {
+		return nil, fmt.Errorf("no NetworkService with name: %v", name)
+	} else {
+		return ns, nil
+	}
+}
+
+func (rc *registryCacheImpl) AddNetworkServiceEndpoint(nse *v1.NetworkServiceEndpoint) (*v1.NetworkServiceEndpoint, error) {
+	nseResponse, err := rc.clientset.NetworkservicemeshV1().NetworkServiceEndpoints("default").Create(nse)
```
Comment (nickolaev) [inline k8s/pkg/registryserver/registry_cache.go]: Are we hardcoding the `default` namespace here?
I know that this is all over the code, but we might want to change this.
@edwarnicke what do you think?

File: k8s/pkg/registryserver/nse.go [hunk 3]
```diff
@@ -125,26 +127,20 @@ func (rs registryService) RemoveNSE(ctx context.Context, request *registry.Remov
...
 
-	lo := metav1.ListOptions{}
-	lo.LabelSelector = "networkservicename=" + request.NetworkServiceName
 	t1 := time.Now()
-	endpointList, e := rs.clientset.Networkservicemesh().NetworkServiceEndpoints("default").List(lo)
-	if e != nil {
-		return nil, e
-	}
+	endpointList := rs.cache.GetNetworkServiceEndpoints(request.NetworkServiceName)
+	logrus.Printf("NSE found %d, retrieve time: %v", len(endpointList), time.Since(t1))
```
Comment (nickolaev) [inline k8s/pkg/registryserver/nse.go]: Please use `Infof` or `Debugf` where appropriate.

File: k8s/pkg/registryserver/registry_cache.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	AddNetworkService(ns *v1.NetworkService) (*v1.NetworkService, error)
+	GetNetworkService(name string) (*v1.NetworkService, error)
+
+	AddNetworkServiceManager(nsm *v1.NetworkServiceManager) (*v1.NetworkServiceManager, error)
+	GetNetworkServiceManager(name string) (*v1.NetworkServiceManager, error)
+
+	AddNetworkServiceEndpoint(nse *v1.NetworkServiceEndpoint) (*v1.NetworkServiceEndpoint, error)
+	DeleteNetworkServiceEndpoint(endpointName string) error
+	GetNetworkServiceEndpoints(networkServiceName string) []*v1.NetworkServiceEndpoint
+
```
Comment (nickolaev) [inline k8s/pkg/registryserver/registry_cache.go]: Hmm can this interface be implemented in a resource neutral way. Something like:

```go
type RegistryCache interface {
		AddEntry(nse interface{}) (interface{}, error)
		DeleteEntryByName(name string) error
		GetEntryByName(name string) []*interface{}
                Start()
                Stop()
}
```

Then each implementation will cast the `interface{}` to whatever it needs to.

**Comment by lobkovilya** [comment]
@nickolaev There is test that checks dynamically adding/deleting resources - [TestNSMDDRegistryNSE](https://github.com/networkservicemesh/networkservicemesh/blob/master/test/integration/nsmd_k8s_test.go)

File: test/integration/nsmd_test_utils/test_utils.go (hunk 1 above)
Comment (lobkovilya) [reply test/integration/nsmd_test_utils/test_utils.go]: Accidentally added these changes to PR, fixed

File: k8s/pkg/registryserver/nse.go (hunk 3 above)
Comment (lobkovilya) [reply k8s/pkg/registryserver/nse.go]: Fixed

File: k8s/pkg/registryserver/registry_cache.go (hunk 4 above)
Comment (lobkovilya) [reply k8s/pkg/registryserver/registry_cache.go]: There is class `abstractResourceCache`. It implements all required logic in resource neutral way. Other classes - `NetworkServiceCache`, `NetworkServiceEndpointCache` and `NetworkServiceManagerCache` are concrete resource caches. They use `abstractResourceCache` inside and do required casts. `RegistryCache` - simple aggregation of all custom resource caches. It starts resource caches in transactional way and delegate Add/Get/Delete methods to appropriate resource cache.
